sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreCheckpointFormatV2Suite.scala

```diff
@@ -0,0 +1,1103 @@
+        }
+        println("wei== res2")
+        res2.foreach(x => for (elem <- x) {
+          println(elem)
+        })
+        idx -> res2
+      }.toMap
+
+      versionToUniqueIdFromCommitLog.foreach {
+        case (version, uniqueIds) =>
+          versionToUniqueIdFromStateStore(version).sameElements(uniqueIds)
+      }
+    }
+  }
+
+  // This test verifies when there are task retries, the unique ids actually stored in
+  // the commit log is the same as those recorded by CkptIdCollectingStateStoreWrapper
+  testWithCheckpointInfoTracked(s"checkpointFormatVersion2 validate ID - task retry") {
+    withTempDir { checkpointDir =>
+      val inputData = MemoryStream[Int]
+      val aggregated =
+        inputData
+          .toDF()
+          .groupBy($"value")
+          .agg(count("*"))
+
+      val writer = (ds: DataFrame, batchId: Long) => {
+        val _ = ds.rdd.filter { x =>
+          val context = TaskContext.get()
+          // Retry in the first attempt
+          if (context.attemptNumber() == 0) {
+            throw new RuntimeException(s"fail the task at " +
+              s"partition ${context.partitionId()} batch Id: $batchId")
+          }
+          x.length >= 0
+        }.collect()
+      }
+
+      val query = aggregated.writeStream
+        .foreachBatch(writer)
+        .option("checkpointLocation", checkpointDir.getAbsolutePath)
+        .outputMode("update")
+        .start()
+
+      inputData.addData(1 to 100)
+      query.processAllAvailable()
+
+      inputData.addData(1 to 100)
+      query.processAllAvailable()
+
+      query.stop()
+
+      val checkpointInfoList = CkptIdCollectingStateStoreWrapper.getStateStoreCheckpointInfos
+      // scalastyle:off line.size.limit
+      // Since every task is retried once, for each partition in each batch, we should have two
+      // state store checkpointInfo. Since these infos are appended sequentially, we can group them
+      // by partitionId and batchVersion, and pick the first one as they are the retried ones.
+      // e.g.
+      // [Picked] StateStoreCheckpointInfo[partitionId=1, batchVersion=2, stateStoreCkptId=Some(a9d5afec-0e8d-4473-b948-6c55513aa509), baseStateStoreCkptId=Some(061f7c53-b300-477a-a599-5387d55e315a)]
+      // [Picked] StateStoreCheckpointInfo[partitionId=0, batchVersion=2, stateStoreCkptId=Some(879cc517-6b85-4dae-abba-794bf2dbab82), baseStateStoreCkptId=Some(513726e7-2448-41a6-a874-92053c5cf86b)]
+      // StateStoreCheckpointInfo[partitionId=1, batchVersion=2, stateStoreCkptId=Some(7f4ad39f-d019-4ca2-8cf4-300379821cd6), baseStateStoreCkptId=Some(061f7c53-b300-477a-a599-5387d55e315a)]
+      // StateStoreCheckpointInfo[partitionId=0, batchVersion=2, stateStoreCkptId=Some(9dc215fe-54f9-4dc1-a59b-a8734f359e46), baseStateStoreCkptId=Some(513726e7-2448-41a6-a874-92053c5cf86b)]
+      // scalastyle:on line.size.limit
+      val pickedCheckpointInfoList = checkpointInfoList
+        .groupBy(x => (x.partitionId, x.batchVersion)).map(_._2.head)
```

Comment:
Does it assume that we always pick the first one in the list? Can we relax the check?

Reply:
relaxed by picking any valid pair